### PR TITLE
feat: Add OIDC multiple audiences support

### PR DESCRIPTION
**Pull request by jmazzitelli**
- Enhanced verifyAudienceClaim() to support multi-audience JWT tokens
- Maintains backward compatibility for single-audience tokens
- Added comprehensive unit tests covering all scenarios
- Added performance benchmarks to validate no regressions
- fixes: #8717

Changes:
- Modified verifyAudienceClaim() to iterate through audience arrays
- Added support for both []string and []interface{} audience formats
- Enhanced error messages for better debugging
- Added 20+ test cases covering edge cases and multi-audience scenarios
- Added performance benchmarks showing <25ns execution time

Performance: Zero memory allocations, <15% performance impact for multi-audience tokens

AI assisted - Cursor / claude-4-sonnet

### CODE REVIEW

**Comment by jmazzitelli** [comment]
[ACCEPTANCE_CRITERIA.md](https://github.com/user-attachments/files/22304537/ACCEPTANCE_CRITERIA.md)
[ARCHITECTURE.md](https://github.com/user-attachments/files/22304538/ARCHITECTURE.md)
[CODEBASE_EXPLORATION.md](https://github.com/user-attachments/files/22304539/CODEBASE_EXPLORATION.md)
[CONSTRAINTS.md](https://github.com/user-attachments/files/22304540/CONSTRAINTS.md)
[DECISIONS.md](https://github.com/user-attachments/files/22304541/DECISIONS.md)
[DOCUMENTATION_INDEX.md](https://github.com/user-attachments/files/22304542/DOCUMENTATION_INDEX.md)
[IMPLEMENTATION_PLAN.md](https://github.com/user-attachments/files/22304543/IMPLEMENTATION_PLAN.md)
[PROBLEM_ANALYSIS.md](https://github.com/user-attachments/files/22304544/PROBLEM_ANALYSIS.md)
[PROGRESS.md](https://github.com/user-attachments/files/22304545/PROGRESS.md)
[REQUIREMENTS.md](https://github.com/user-attachments/files/22304546/REQUIREMENTS.md)
[SESSION_STATUS.md](https://github.com/user-attachments/files/22304547/SESSION_STATUS.md)
[SOLUTION_EVALUATION.md](https://github.com/user-attachments/files/22304548/SOLUTION_EVALUATION.md)
[SOLUTION_OPTIONS.md](https://github.com/user-attachments/files/22304549/SOLUTION_OPTIONS.md)
[TECHNICAL_SPEC.md](https://github.com/user-attachments/files/22304550/TECHNICAL_SPEC.md)
[TEST_STRATEGY.md](https://github.com/user-attachments/files/22304551/TEST_STRATEGY.md)
[USER_STORIES.md](https://github.com/user-attachments/files/22304553/USER_STORIES.md)

File: handlers/authentication/openid_auth_controller.go [hunk 1]
```diff
@@ -1247,24 +1247,39 @@ func verifyAudienceClaim(openIdParams *openidFlowHelper, oidCfg config.OpenIdCon
 	} else {
 		switch ac := audienceClaim.(type) {
 		case string:
+			// Existing single-audience validation (unchanged)
```
Comment (nrfox) [inline handlers/authentication/openid_auth_controller.go]: You can remove these AI generated comments. They're not helpful.

File: handlers/authentication/openid_auth_controller.go [hunk 2]
```diff
@@ -1247,24 +1247,39 @@ func verifyAudienceClaim(openIdParams *openidFlowHelper, oidCfg config.OpenIdCon
...
+			if len(audArray) == 0 {
+				return errors.New("the OpenId token contains empty audience list")
 			}
-			acStr := fmt.Sprintf("%v", audienceClaim.([]any)[0])
-			if oidCfg.ClientId != acStr {
-				return fmt.Errorf("the OpenId token is not targeted for Kiali; got []aud = '%v'", acStr)
+
+			// Check if ClientId is found in any of the audiences
+			for _, audItem := range audArray {
+				if audStr, ok := audItem.(string); ok {
```
Comment (nrfox) [inline handlers/authentication/openid_auth_controller.go]: This is a different assertion than what was there previously. In the old logic, we convert to string with `fmt.Sprintf`. In the new logic, we type assert that the item is a string which is less flexible than converting to a string.

File: handlers/authentication/openid_auth_controller_benchmark_test.go
Comment (nrfox) [inline handlers/authentication/openid_auth_controller_benchmark_test.go:1]: We don't need these benchmark tests. If the audience array was huge then it might impact performance but that is unlikely and there's probably limitations on how large that field can be in the spec.

File: handlers/authentication/openid_auth_controller_test.go [hunk 3]
```diff
@@ -51,29 +51,136 @@ func TestVerifyAudienceClaim(t *testing.T) {
...
+		err := verifyAudienceClaim(&oip, oidCfg)
+		assert.Nil(t, err, "verifyAudienceClaim should pass when valid audience is found")
+	})
+
+	t.Run("NonStringInArray", func(t *testing.T) {
+		oip.IdTokenPayload["aud"] = []interface{}{123, 456}
+		err := verifyAudienceClaim(&oip, oidCfg)
+		assert.NotNil(t, err, "verifyAudienceClaim should have failed")
+		assert.Contains(t, err.Error(), "not found in audiences")
+	})
```
Comment (nrfox) [inline handlers/authentication/openid_auth_controller_test.go:169]: Technically I think this was possible in Kiali before but the spec does expect this to be strings:
> 	In the general case, the aud value is an array of case-sensitive strings. In the common special case when there is one audience, the aud value MAY be a single case-sensitive string. 

From the [spec](https://openid.net/specs/openid-connect-core-1_0.html)

File: handlers/authentication/openid_auth_controller_test.go [hunk 4]
```diff
@@ -51,29 +51,136 @@ func TestVerifyAudienceClaim(t *testing.T) {
...
+		assert.NotNil(t, err, "verifyAudienceClaim should have failed")
+		assert.Contains(t, err.Error(), "has no aud claim")
+	})
+
+	t.Run("InvalidType", func(t *testing.T) {
+		oip.IdTokenPayload["aud"] = 12345
+		err := verifyAudienceClaim(&oip, oidCfg)
+		assert.NotNil(t, err, "verifyAudienceClaim should have failed")
+		assert.Contains(t, err.Error(), "unexpected audience claim")
+	})
```
Comment (nrfox) [inline handlers/authentication/openid_auth_controller_test.go]: This looks like a dupe of `NonStringInArray`?

File: handlers/authentication/openid_auth_controller.go (hunk 1 above)
Comment (jmazzitelli) [reply handlers/authentication/openid_auth_controller.go]: i'm re-wording some comments, will remove others.

File: handlers/authentication/openid_auth_controller_benchmark_test.go
Comment (jmazzitelli) [reply handlers/authentication/openid_auth_controller_benchmark_test.go:1]: that benchmark file will be removed

File: handlers/authentication/openid_auth_controller.go (hunk 2 above)
Comment (jmazzitelli) [reply handlers/authentication/openid_auth_controller.go]: this will be brought back to the way it was , using Sprintf to convert to string.

File: handlers/authentication/openid_auth_controller_test.go (hunk 3 above)
Comment (jmazzitelli) [reply handlers/authentication/openid_auth_controller_test.go:169]: The code and tests now allow for numeric audiences, even if spec says that's not the general case.

File: handlers/authentication/openid_auth_controller_test.go (hunk 4 above)
Comment (jmazzitelli) [reply handlers/authentication/openid_auth_controller_test.go]: not a dup. it was checking the default case (audience as a single numeric value) whereas NonStringInArray was checking interface array. But I changed the default case to allow for single numeric value as an audience, just convert to string as the array case does. unit tests now cover this.

**Comment by jmazzitelli** [comment]
not sure what's up with the failures - doesn't look related to this PR

**Comment by jmazzitelli** [comment]
## AI Development Context Files

This PR includes comprehensive AI-assisted development documentation. The attached zip file contains:

- **Architecture & Design**: Technical specifications, architecture decisions, and solution evaluations
- **Implementation Details**: Step-by-step implementation plans and reviews  
- **Testing Strategy**: Comprehensive test plans, results, and performance reports
- **Session Documentation**: Progress tracking, status reports, and session summaries
- **Requirements & Constraints**: User stories, acceptance criteria, and technical constraints

**File**: `ai-dev-context.zip` (161KB) - Contains 40+ documentation files covering the entire development lifecycle for the OIDC multiple audiences feature.

[ai-dev-context.zip](https://github.com/user-attachments/files/22451509/ai-dev-context.zip)


The documentation provides complete context for understanding the implementation approach, testing strategy, and architectural decisions made during development.

**Review by nrfox — COMMENTED** [review]
The auth controller LGTM. I have not tested the hack scripts. Why was dex replaced with hydra?

**Comment by jmazzitelli** [comment]
> Why was dex replaced with hydra?

The version of DEX we were using apparently did not support multiple audiences in the manner we needed.
